src/coreclr/jit/morph.cpp

```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
     return op1;
 }
 
+//------------------------------------------------------------------------------
+// fgMorphOpCast : Try to simplify "(T)x op (T)y" to "(T)(x op y)".
+// https://github.com/dotnet/runtime/issues/13816
+//
+// Arguments:
+//       tree - node to fold
+//
+// return value:
+//       A folded GenTree* instance, or the same GenTree* if it couldn't be folded
+//
+
+GenTree* Compiler::fgMorphOpCast(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree->TypeGet()));
+    assert(tree->OperIs(GT_OR, GT_AND, GT_XOR));
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+    genTreeOps oper = tree->OperGet();
+
+    // see whether both ops are casts, with matching to and from types
+    if (op1->OperIs(GT_CAST) && op2->OperIs(GT_CAST))
+    {
+        var_types fromType = op1->CastFromType();
+        var_types toType = op1->CastToType();
+        GenTreeFlags gtf_unsigned = op1->gtFlags & GTF_UNSIGNED;
+
+        if (op2->CastFromType() != fromType || op2->CastToType() != toType || (op2->gtFlags & GTF_UNSIGNED) != gtf_unsigned)
+        {
+            return tree;
+        }
+
+        //     tree             CAST
+        //     /   \             |
+        //   CAST  CAST   ==>   tree
+        //    |     |          /   \
+        //   op1   op2        op1  op2
+
+        tree->gtOp1 = op1->AsCast()->CastOp();
+        tree->gtOp2 = op2->AsCast()->CastOp();
+        tree->gtType = fromType;
+
+        GenTree* result = gtNewCastNode(toType, tree, gtf_unsigned != 0, toType);
```

Comment:
```suggestion
        GenTree* result = gtNewCastNode(genActualType(toType), tree, gtf_unsigned != 0, toType);
```
Cast nodes always have the normalized stack type (and in this case `toType` can be a small type so it "matters").

Throughout nit: you could reuse one of the cast nodes here, to improve the density of IR.

Reply:
Apologies, I'm still learning - do you happen to know of somewhere I can read about "normalised stack type" and "small type"?

Reply:
Of course: look in the `vartypes.h` header - there exists a mapping between all the types and their "actual" versions (aka on-the-IL stack ones). We call integers less than 4 bytes in size (so, `TYP_BYTE`, `TYP_BOOL`, `TYP_UBYTE`, `TYP_USHORT`, `TYP_SHORT`) "small". They usually require special considerations when writing code (see also my [notes](https://gist.github.com/SingleAccretion/d6033fc5ada61955e868b8cbfb2bd822) about them).

Reply:
Thanks!

cdec019d463276d2546e47e51d6894e951f0a7b7